File: android/app/src/main/kotlin/com/example/mobile_offline/FaceLandmarkerBridge.kt
package com.example.mobile_offline

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarkerResult
import java.io.ByteArrayInputStream

class FaceLandmarkerBridge(private val context: Context) {
    private var faceLandmarker: FaceLandmarker? = null
    private val processLock = Any()

    fun isInitialized(): Boolean = synchronized(processLock) { faceLandmarker != null }

    fun initialize(modelAssetPath: String = "face_landmarker.task") {
        synchronized(processLock) {
            if (faceLandmarker != null) return

            // Load MediaPipe vision task classes on this thread before any background work.
            // BaseVisionTaskApi.<clinit> SIGSEGVs on Android 16 if first touched off-main.
            Class.forName("com.google.mediapipe.tasks.vision.core.BaseVisionTaskApi")

            // Fail fast with a clear message if the model asset is missing.
            context.assets.open(modelAssetPath).use { }

            val baseOptions = BaseOptions.builder()
                .setModelAssetPath(modelAssetPath)
                .build()

            val options = FaceLandmarker.FaceLandmarkerOptions.builder()
                .setBaseOptions(baseOptions)
                .setRunningMode(RunningMode.IMAGE)
                .setNumFaces(1)
                .setMinFaceDetectionConfidence(0.5f)
                .setMinFacePresenceConfidence(0.5f)
                .setMinTrackingConfidence(0.5f)
                .setOutputFaceBlendshapes(true)
                .build()
            faceLandmarker = FaceLandmarker.createFromOptions(context, options)
        }
    }

    fun processFrame(frameBytes: ByteArray): Map<String, Any> {
        if (frameBytes.isEmpty()) {
            return emptyFaceResult()
        }

        synchronized(processLock) {
            val landmarker = faceLandmarker ?: return emptyFaceResult()

            var bitmap: Bitmap? = null
            try {
                val decodeOptions = BitmapFactory.Options().apply {
                    inPreferredConfig = Bitmap.Config.ARGB_8888
                }
                bitmap = BitmapFactory.decodeStream(ByteArrayInputStream(frameBytes), null, decodeOptions)
                    ?: return emptyFaceResult()
                if (bitmap.width <= 0 || bitmap.height <= 0) return emptyFaceResult()

                val softwareBitmap = ensureSoftwareBitmap(bitmap)
                if (softwareBitmap !== bitmap) {
                    bitmap.recycle()
                    bitmap = softwareBitmap
                }

                val mpImage = BitmapImageBuilder(softwareBitmap).build()
                val result: FaceLandmarkerResult = landmarker.detect(mpImage)
                if (result.faceLandmarks().isEmpty()) {
                    return emptyFaceResult()
                }

                val blendshapes = extractBlendshapes(result)
                val mouthOpen = blendshapes["jawOpen"] ?: 0.0
                val mouthPucker = blendshapes["mouthPucker"] ?: 0.0
                val smileLeft = blendshapes["mouthSmileLeft"] ?: 0.0
                val smileRight = blendshapes["mouthSmileRight"] ?: 0.0
                val smile = (smileLeft + smileRight) / 2.0
                val mouthClose = blendshapes["mouthClose"] ?: 0.0
                val mouthFunnel = blendshapes["mouthFunnel"] ?: 0.0
                val stretchLeft = blendshapes["mouthStretchLeft"] ?: 0.0
                val stretchRight = blendshapes["mouthStretchRight"] ?: 0.0
                val mouthStretch = (stretchLeft + stretchRight) / 2.0

                val mouthBox = mouthBoundingBox(result)

                return mapOf(
                    "faceDetected" to true,
                    "mouthOpen" to mouthOpen,
                    "mouthPucker" to mouthPucker,
                    "smile" to smile,
                    "mouthClose" to mouthClose,
                    "mouthFunnel" to mouthFunnel,
                    "mouthStretch" to mouthStretch,
                    "mouthMinX" to mouthBox[0],
                    "mouthMinY" to mouthBox[1],
                    "mouthMaxX" to mouthBox[2],
                    "mouthMaxY" to mouthBox[3],
                    "ts" to System.currentTimeMillis()
                )
            } catch (_: Exception) {
                return emptyFaceResult()
            } finally {
                bitmap?.recycle()
            }
        }
    }

    fun close() {
        synchronized(processLock) {
            faceLandmarker?.close()
            faceLandmarker = null
        }
    }

    private fun extractBlendshapes(result: FaceLandmarkerResult): Map<String, Double> {
        val out = HashMap<String, Double>()
        val optional = result.faceBlendshapes()
        if (!optional.isPresent) return out
        val faces = optional.get()
        if (faces.isEmpty()) return out
        for (category in faces[0]) {
            val name = category.categoryName() ?: continue
            out[name] = category.score().toDouble()
        }
        return out
    }

    /** Tight outer-lip box: corners + upper/lower mid (avoids tall multi-ring span). */
    private fun mouthBoundingBox(result: FaceLandmarkerResult): DoubleArray {
        val landmarks = result.faceLandmarks().firstOrNull() ?: return doubleArrayOf(0.0, 0.0, 0.0, 0.0)
        // 61/291 = mouth corners; 0/17 = outer upper/lower lip; 13/14 = mid lip
        val lipIndices = intArrayOf(61, 291, 0, 17, 13, 14)
        var minX = 1.0
        var minY = 1.0
        var maxX = 0.0
        var maxY = 0.0
        var any = false
        for (idx in lipIndices) {
            if (idx < 0 || idx >= landmarks.size) continue
            val lm = landmarks[idx]
            val x = lm.x().toDouble()
            val y = lm.y().toDouble()
            if (x < minX) minX = x
            if (y < minY) minY = y
            if (x > maxX) maxX = x
            if (y > maxY) maxY = y
            any = true
        }
        return if (any) doubleArrayOf(minX, minY, maxX, maxY) else doubleArrayOf(0.0, 0.0, 0.0, 0.0)
    }

    private fun emptyFaceResult(): Map<String, Any> = mapOf(
        "faceDetected" to false,
        "mouthOpen" to 0.0,
        "mouthPucker" to 0.0,
        "smile" to 0.0,
        "mouthClose" to 0.0,
        "mouthFunnel" to 0.0,
        "mouthStretch" to 0.0,
        "mouthMinX" to 0.0,
        "mouthMinY" to 0.0,
        "mouthMaxX" to 0.0,
        "mouthMaxY" to 0.0,
        "ts" to System.currentTimeMillis()
    )

    private fun ensureSoftwareBitmap(source: Bitmap): Bitmap {
        if (source.config == Bitmap.Config.ARGB_8888 && !source.isRecycled) {
            return source
        }
        return source.copy(Bitmap.Config.ARGB_8888, false) ?: source
    }
}

File: ios/Runner/FaceLandmarkerBridge.swift
import Foundation
import MediaPipeTasksVision
import UIKit

final class FaceLandmarkerBridge {
  private var faceLandmarker: FaceLandmarker?

  func isInitialized() -> Bool { faceLandmarker != nil }

  func initialize(modelPath: String) throws {
    if faceLandmarker != nil { return }

    let baseOptions = BaseOptions(modelAssetPath: modelPath)
    var options = FaceLandmarkerOptions()
    options.baseOptions = baseOptions
    options.runningMode = .image
    options.numFaces = 1
    options.minFaceDetectionConfidence = 0.5
    options.minFacePresenceConfidence = 0.5
    options.minTrackingConfidence = 0.5
    options.outputFaceBlendshapes = true
    faceLandmarker = try FaceLandmarker(options: options)
  }

  func processFrame(frameBytes: Data) -> [String: Any] {
    guard !frameBytes.isEmpty else { return emptyFaceResult() }
    guard let faceLandmarker else { return emptyFaceResult() }
    guard let image = UIImage(data: frameBytes), let cgImage = image.cgImage else {
      return emptyFaceResult()
    }
    guard cgImage.width > 0, cgImage.height > 0 else { return emptyFaceResult() }

    let mpImage = try? MPImage(uiImage: UIImage(cgImage: cgImage))
    guard let mpImage else { return emptyFaceResult() }
    guard let result = try? faceLandmarker.detect(image: mpImage) else { return emptyFaceResult() }
    guard let face = result.faceLandmarks.first else { return emptyFaceResult() }

    let blendshapes = extractBlendshapes(from: result)
    let mouthOpen = blendshapes["jawOpen"] ?? 0
    let mouthPucker = blendshapes["mouthPucker"] ?? 0
    let smileLeft = blendshapes["mouthSmileLeft"] ?? 0
    let smileRight = blendshapes["mouthSmileRight"] ?? 0
    let smile = (smileLeft + smileRight) / 2.0
    let mouthClose = blendshapes["mouthClose"] ?? 0
    let mouthFunnel = blendshapes["mouthFunnel"] ?? 0
    let stretchLeft = blendshapes["mouthStretchLeft"] ?? 0
    let stretchRight = blendshapes["mouthStretchRight"] ?? 0
    let mouthStretch = (stretchLeft + stretchRight) / 2.0
    let box = mouthBoundingBox(landmarks: face)

    return [
      "faceDetected": true,
      "mouthOpen": mouthOpen,
      "mouthPucker": mouthPucker,
      "smile": smile,
      "mouthClose": mouthClose,
      "mouthFunnel": mouthFunnel,
      "mouthStretch": mouthStretch,
      "mouthMinX": box.0,
      "mouthMinY": box.1,
      "mouthMaxX": box.2,
      "mouthMaxY": box.3,
      "ts": Int(Date().timeIntervalSince1970 * 1000)
    ]
  }

  private func extractBlendshapes(from result: FaceLandmarkerResult) -> [String: Double] {
    var out: [String: Double] = [:]
    guard let classifications = result.faceBlendshapes.first else { return out }
    for category in classifications.categories {
      let name = category.categoryName
      guard !name.isEmpty else { continue }
      out[name] = Double(category.score)
    }
    return out
  }

  private func mouthBoundingBox(landmarks: [NormalizedLandmark]) -> (Double, Double, Double, Double) {
    // Tight outer-lip box: 61/291 corners, 0/17 vertical, 13/14 mid lip
    let lipIndices = [61, 291, 0, 17, 13, 14]
    var minX = 1.0
    var minY = 1.0
    var maxX = 0.0
    var maxY = 0.0
    var any = false
    for idx in lipIndices {
      guard idx >= 0, idx < landmarks.count else { continue }
      let x = Double(landmarks[idx].x)
      let y = Double(landmarks[idx].y)
      minX = min(minX, x)
      minY = min(minY, y)
      maxX = max(maxX, x)
      maxY = max(maxY, y)
      any = true
    }
    return any ? (minX, minY, maxX, maxY) : (0, 0, 0, 0)
  }

  private func emptyFaceResult() -> [String: Any] {
    [
      "faceDetected": false,
      "mouthOpen": 0.0,
      "mouthPucker": 0.0,
      "smile": 0.0,
      "mouthClose": 0.0,
      "mouthFunnel": 0.0,
      "mouthStretch": 0.0,
      "mouthMinX": 0.0,
      "mouthMinY": 0.0,
      "mouthMaxX": 0.0,
      "mouthMaxY": 0.0,
      "ts": Int(Date().timeIntervalSince1970 * 1000)
    ]
  }
}
